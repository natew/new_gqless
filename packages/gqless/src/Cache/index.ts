import lodashGet from 'lodash/get';
import lodashMerge from 'lodash/merge';

import { Selection } from '../Selection';

export const CacheNotFound = Symbol('Not Found');

export function createCache() {
  const cache: Record<string, unknown> = {};

  function getCacheFromSelection(selection: Selection) {
    const path = selection.path.slice(1);
    return lodashGet(cache, path, CacheNotFound);
  }

  function mergeCache(data: unknown) {
    lodashMerge(cache, data);
  }

  return { cache, getCacheFromSelection, mergeCache };
}
