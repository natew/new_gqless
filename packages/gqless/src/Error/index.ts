import type { GraphQLError } from 'graphql';

export class gqlessError extends Error {
  graphQLErrors?: ReadonlyArray<GraphQLError>;
  otherError?: unknown;

  constructor(
    message: string,
    {
      graphQLErrors,
      otherError,
      caller,
    }: {
      graphQLErrors?: gqlessError['graphQLErrors'];
      otherError?: gqlessError['otherError'];
      caller?: Function;
    } = {}
  ) {
    super(message);

    if (graphQLErrors) this.graphQLErrors = graphQLErrors;
    if (otherError !== undefined) this.otherError = otherError;

    /* istanbul ignore else */
    if (caller && Error.captureStackTrace!)
      Error.captureStackTrace(this, caller);
  }

  toJSON() {
    return {
      message: this.message,
      graphQLErrors: this.graphQLErrors,
      otherError: this.otherError,
    };
  }

  static create(error: unknown, caller?: Function): gqlessError {
    let newError: gqlessError;

    if (error instanceof gqlessError) newError = error;
    else if (error instanceof Error)
      newError = Object.assign(new gqlessError(error.message), error);
    else
      newError = new gqlessError('Unexpected error type', {
        otherError: error,
      });

    /* istanbul ignore else */
    if (caller && Error.captureStackTrace!)
      Error.captureStackTrace(newError, caller);

    return newError;
  }
}

export * from './retry';
