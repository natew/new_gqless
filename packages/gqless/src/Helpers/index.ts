export { selectFields } from './selectFields';
export type { HydrateCacheOptions } from './ssr';
export * from './getFields';
