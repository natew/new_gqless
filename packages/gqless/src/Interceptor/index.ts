import { Selection } from '../Selection';

export class Interceptor {
  fetchSelections = new Set<Selection>();
  listening = true;
  selectionAddListeners = new Set<(selection: Selection) => void>();
  selectionCacheListeners = new Set<(selection: Selection) => void>();
  selectionCacheRefetchListeners = new Set<(selection: Selection) => void>();

  addSelection(selection: Selection) {
    if (this.listening) {
      this.fetchSelections.add(selection);

      for (const listener of this.selectionAddListeners) {
        listener(selection);
      }
    }
  }

  addSelectionCache(selection: Selection) {
    if (this.listening) {
      for (const listener of this.selectionCacheListeners) {
        listener(selection);
      }
    }
  }

  addSelectionCacheRefetch(selection: Selection) {
    if (this.listening && this.selectionCacheRefetchListeners.size) {
      this.fetchSelections.add(selection);

      for (const listener of this.selectionCacheRefetchListeners) {
        listener(selection);
      }
    }
  }

  removeSelections(selections: Set<Selection> | Selection[]) {
    for (const selection of selections) {
      this.fetchSelections.delete(selection);
    }
  }
}

export type InterceptorManager = ReturnType<typeof createInterceptorManager>;

export function createInterceptorManager() {
  const interceptors = new Set<Interceptor>();

  const globalInterceptor = new Interceptor();
  interceptors.add(globalInterceptor);

  function createInterceptor() {
    const interceptor = new Interceptor();
    interceptors.add(interceptor);
    return interceptor;
  }

  function removeInterceptor(interceptor: Interceptor) {
    interceptors.delete(interceptor);
  }

  function addSelection(selection: Selection) {
    for (const interceptor of interceptors) {
      interceptor.addSelection(selection);
    }
  }

  function addSelectionCache(selection: Selection) {
    for (const interceptor of interceptors) {
      if (interceptor === globalInterceptor) continue;

      interceptor.addSelectionCache(selection);
    }
  }

  function addSelectionCacheRefetch(selection: Selection) {
    for (const interceptor of interceptors) {
      if (interceptor === globalInterceptor) continue;

      interceptor.addSelectionCacheRefetch(selection);
    }
  }

  function addSelections(selection: Selection[] | Set<Selection>) {
    selection.forEach(addSelection);
  }

  function removeSelections(selections: Selection[] | Set<Selection>) {
    for (const interceptor of interceptors) {
      interceptor.removeSelections(selections);
    }
  }

  return {
    interceptors,
    globalInterceptor,
    createInterceptor,
    removeInterceptor,
    addSelection,
    addSelectionCache,
    addSelectionCacheRefetch,
    addSelections,
    removeSelections,
  };
}
