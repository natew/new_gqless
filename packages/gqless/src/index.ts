export * from './Client';
export * from './Selection';
export * from './Schema';
export * from './Error';
export * from './Poller';
export * from './Helpers';
export * from './Utils/cycle';
