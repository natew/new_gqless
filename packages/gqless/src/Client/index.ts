import { CacheNotFound, createCache } from "../Cache";
import { buildQuery } from "../QueryBuilder";
import { Selection } from "../Selection/selection";
import { QueryFetcher, ScalarsHash, Schema } from "../types";
import { deferredPromise } from "../Utils/promise";

const ProxySymbol = Symbol("gqless-proxy");

export function createClient<GeneratedSchema = never>(
  schema: Readonly<Schema>,
  scalars: ScalarsHash,
  queryFetcher: QueryFetcher
) {
  const ProxyCache = new WeakMap<Selection, object>();
  const ProxyCacheReverse = new WeakMap<object, Selection>();

  const ProxyResolvePromises = new WeakMap<object, Promise<void>>();

  const globalSelections = new Set<Selection>();
  const client: GeneratedSchema = createSchemaProxy();
  const { getCacheFromSelection, mergeCache } = createCache();

  function resolve(proxy: object): Promise<void> {
    return (
      ProxyResolvePromises.get(proxy) ||
      (() => {
        const proxySelection = ProxyCacheReverse.get(proxy);
        if (!proxySelection) throw Error("Couldn't find any selection with the proxy");

        const selections = [...globalSelections].filter((selection) => {
          return selection.selections.has(selection);
        });

        if (selections.length === 0) {
          console.warn("No selections made");
          return Promise.resolve();
        }

        return resolveSelections(selections);
      })()
    );
  }

  async function resolveSelections(selections: Selection[] | Set<Selection>) {
    const { reject, resolve, promise } = deferredPromise();
    for (const selection of selections) {
      const proxySelection = ProxyCache.get(selection);
      if (proxySelection) {
        ProxyResolvePromises.set(selection, promise);
      }
    }

    try {
      const { query, variables } = buildQuery(selections);

      const { data, errors } = await queryFetcher(query, variables);

      if (data) {
        mergeCache(data);

        for (const selection of selections) {
          globalSelections.delete(selection);
        }
      }

      if (errors) {
        console.error(errors);
        const err = Error("Errors in resolve");

        if (Error.captureStackTrace) {
          Error.captureStackTrace(err, resolveAllSelections);
        }

        throw err;
      }

      resolve();
    } catch (err) {
      reject(err);
      throw err;
    }
  }

  async function resolveAllSelections() {
    const selections = [...globalSelections];
    const { reject, resolve, promise } = deferredPromise();
    for (const selection of selections) {
      const proxySelection = ProxyCache.get(selection);
      if (proxySelection) {
        ProxyResolvePromises.set(selection, promise);
      }
    }

    try {
      await resolveSelections(selections);

      resolve();
    } catch (err) {
      reject(err);
      throw err;
    }
  }

  // TODO: Refetch behavior
  function createArrayTypeProxy(schemaType: Schema[string], selectionsArg: Selection) {
    const arrayCacheValue = getCacheFromSelection(selectionsArg);
    if (arrayCacheValue === null) return null;

    const proxy =
      ProxyCache.get(selectionsArg) ||
      (() => {
        const createdProxy: ProxyConstructor = new Proxy(
          arrayCacheValue === CacheNotFound ? [ProxySymbol] : arrayCacheValue,
          {
            get(target, key: string, receiver) {
              const index = parseInt(key);

              if (Number.isInteger(index)) {
                if (arrayCacheValue !== CacheNotFound && arrayCacheValue[index] == null) {
                  // if arrayCacheValue[index] is 'null' or 'undefined'
                  return arrayCacheValue[index];
                }

                const selection = new Selection({
                  key: index,
                  prevSelection: selectionsArg,
                });
                return createTypeProxy(schemaType, selection);
              }
              return Reflect.get(target, key, receiver);
            },
          }
        );
        ProxyCache.set(selectionsArg, createdProxy);
        ProxyCacheReverse.set(createdProxy, selectionsArg);
        return createdProxy;
      })();

    return proxy;
  }

  // TODO: Refetch behavior
  function createTypeProxy(schemaType: Schema[string], selectionsArg: Selection) {
    const cacheValue = getCacheFromSelection(selectionsArg);
    if (cacheValue === null) return null;

    const proxy =
      ProxyCache.get(selectionsArg) ||
      (() => {
        const createdProxy = new Proxy(
          Object.fromEntries(
            Object.keys(schemaType).map((key) => {
              return [key, ProxySymbol];
            })
          ) as Record<string, unknown>,
          {
            get(target, key: string, receiver) {
              if (!schemaType.hasOwnProperty(key)) return Reflect.get(target, key, receiver);

              const value = schemaType[key];
              if (value) {
                const { __type, __args } = value;

                const { pureType, isArray } = (() => {
                  let isNullable = true;
                  let nullableItems = true;
                  let isArray = false;
                  let pureType = __type;
                  if (__type.endsWith("!")) {
                    isNullable = false;
                    pureType = __type.slice(0, __type.length - 1);
                  }
                  if (pureType.startsWith("[")) {
                    pureType = pureType.slice(1, pureType.length - 1);
                    isArray = true;
                    if (pureType.endsWith("!")) {
                      nullableItems = false;
                      pureType = pureType.slice(0, pureType.length - 1);
                    }
                  }

                  return {
                    pureType,
                    isNullable,
                    nullableItems,
                    isArray,
                  };
                })();

                const selection = new Selection({
                  key,
                  prevSelection: selectionsArg,
                  isArray,
                });

                const resolve = (): unknown => {
                  if (scalars[pureType]) {
                    const cacheValue = getCacheFromSelection(selection);

                    if (cacheValue === CacheNotFound) {
                      globalSelections.add(selection);

                      return null;
                    }

                    return cacheValue;
                  }

                  const typeValue = schema[pureType];
                  if (typeValue) {
                    if (isArray) {
                      return createArrayTypeProxy(typeValue, selection);
                    }
                    return createTypeProxy(typeValue, selection);
                  }

                  throw Error("97 Not found!");
                };

                if (__args) {
                  return (args: typeof __args) => {
                    selection.args = args;
                    selection.argTypes = __args;

                    return resolve();
                  };
                }

                return resolve();
              }
              console.error("Not found", key);

              throw Error(`83. Not found`);
            },
          }
        );
        ProxyCache.set(selectionsArg, createdProxy);
        ProxyCacheReverse.set(createdProxy, selectionsArg);
        return createdProxy;
      })();

    return proxy;
  }

  function createSchemaProxy() {
    return new Proxy(
      Object.fromEntries(
        Object.keys(schema).map((key) => {
          return [key, ProxySymbol];
        })
      ) as Record<string, unknown>,
      {
        get(_target, key: string, _receiver) {
          const value = schema[key];

          if (value) {
            const selection = new Selection({
              key,
            });

            return createTypeProxy(value, selection);
          }
          throw Error("104. Not found");
        },
      }
    ) as GeneratedSchema;
  }

  return {
    client,
    globalSelections,
    resolveAllSelections,
    resolve,
  };
}
