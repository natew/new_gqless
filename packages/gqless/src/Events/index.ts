import { ExecutionResult } from 'graphql';

import { CacheType } from '../Cache';
import { gqlessError } from '../Error';
import { Selection } from '../Selection';

export interface FetchEventData {
  executionResult?: Pick<ExecutionResult, 'data' | 'extensions'>;
  error?: gqlessError;
  query: string;
  variables: Record<string, unknown> | undefined;
  cacheSnapshot: CacheType;
  selections: Selection[];
  type: 'query' | 'mutation' | 'subscription';
}

export interface CacheChangeEventData {
  selection: Selection;
  data: unknown;
}

interface OnFetchEventFn {
  (data: Promise<FetchEventData>): void;
}

interface OnCacheChangeEventFn {
  (data: CacheChangeEventData): void;
}

export class EventHandler {
  public hasFetchSubscribers = false;
  private onFetchListeners = new Set<OnFetchEventFn>();

  private onCacheChangeListeners = new Set<OnCacheChangeEventFn>();

  public sendCacheChange(data: CacheChangeEventData) {
    this.onCacheChangeListeners.forEach((listener) => {
      listener(data);
    });
  }

  public sendFetchPromise(data: Promise<FetchEventData>) {
    this.onFetchListeners.forEach((listener) => {
      listener(data);
    });
  }

  public onCacheChangeSubscribe(fn: OnCacheChangeEventFn) {
    const self = this;

    self.onCacheChangeListeners.add(fn);

    return function unsubscribe() {
      self.onCacheChangeListeners.delete(fn);
    };
  }

  public onFetchSubscribe(fn: OnFetchEventFn) {
    const self = this;

    self.hasFetchSubscribers = Boolean(self.onFetchListeners.add(fn).size);

    return function unsubscribe() {
      self.onFetchListeners.delete(fn);
      self.hasFetchSubscribers = Boolean(self.onFetchListeners.size);
    };
  }
}
