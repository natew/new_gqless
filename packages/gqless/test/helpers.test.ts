import { selectFields } from '../src';
import { createTestClient } from './utils';

describe('selectFields', () => {
  test('recursive *, depth 1', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(
        query.human({
          name: 'foo',
        })
      );
    });

    expect(data).toEqual({
      __typename: 'Human',
      name: 'foo',
      father: null,
      nullFather: null,
      sons: [null],
    });
  });

  test('recursive *, depth 2', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(
        query.human({
          name: 'foo',
        }),
        '*',
        2
      );
    });

    expect(data).toEqual({
      __typename: 'Human',
      name: 'foo',
      father: {
        __typename: 'Human',
        name: 'default',
        father: null,
        nullFather: null,
        sons: [null],
      },
      nullFather: null,
      sons: [
        {
          __typename: 'Human',
          name: 'default',
          father: null,
          nullFather: null,
          sons: [null],
        },
        {
          __typename: 'Human',
          name: 'default',
          father: null,
          nullFather: null,
          sons: [null],
        },
      ],
    });
  });

  test('named no recursive', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(
        query.human({
          name: 'bar',
        }),
        ['name', 'father.name']
      );
    });

    expect(data).toEqual({
      name: 'bar',
      father: {
        name: 'default',
      },
    });
  });

  test('named recursive, depth 1', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(
        query.human({
          name: 'bar',
        }),
        ['father']
      );
    });

    expect(data).toEqual({
      father: {
        __typename: 'Human',
        name: 'default',
        father: null,
        nullFather: null,
        sons: [null],
      },
    });
  });

  test('named recursive, depth 2', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(
        query.human({
          name: 'bar',
        }),
        ['father'],
        2
      );
    });

    expect(data).toEqual({
      father: {
        __typename: 'Human',
        name: 'default',
        father: {
          __typename: 'Human',
          name: 'default',
          father: null,
          sons: [null],
          nullFather: null,
        },
        nullFather: null,
        sons: [
          {
            __typename: 'Human',
            name: 'default',
            father: null,
            nullFather: null,
            sons: [null],
          },
          {
            __typename: 'Human',
            name: 'default',
            father: null,
            nullFather: null,
            sons: [null],
          },
        ],
      },
    });
  });

  test('named recursive - array', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human().sons, ['name']);
    });

    expect(data).toEqual([
      {
        name: 'default',
      },
      {
        name: 'default',
      },
    ]);
  });

  test('recursive * - array', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human().sons, '*');
    });

    expect(data).toEqual([
      {
        __typename: 'Human',
        father: null,
        nullFather: null,
        sons: [null],
        name: 'default',
      },
      {
        __typename: 'Human',
        father: null,
        nullFather: null,
        sons: [null],
        name: 'default',
      },
    ]);
  });

  test('empty named fields array', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human(), []);
    });

    expect(data).toEqual({});
  });

  test('named fields array values - depth 1', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human(), ['sons']);
    });

    expect(data).toEqual({
      sons: [
        {
          __typename: 'Human',
          name: 'default',
          father: null,
          nullFather: null,
          sons: [null],
        },
        {
          __typename: 'Human',
          name: 'default',
          father: null,
          nullFather: null,
          sons: [null],
        },
      ],
    });
  });

  test('named fields array values - depth 2', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human(), ['sons'], 2);
    });

    expect(data).toEqual({
      sons: [
        {
          __typename: 'Human',
          name: 'default',
          father: {
            __typename: 'Human',
            name: 'default',
            father: null,
            nullFather: null,
            sons: [null],
          },
          nullFather: null,
          sons: [
            {
              __typename: 'Human',
              name: 'default',
              father: null,
              nullFather: null,
              sons: [null],
            },
            {
              __typename: 'Human',
              name: 'default',
              father: null,
              nullFather: null,
              sons: [null],
            },
          ],
        },
        {
          __typename: 'Human',
          name: 'default',
          father: {
            __typename: 'Human',
            name: 'default',
            father: null,
            nullFather: null,
            sons: [null],
          },
          nullFather: null,
          sons: [
            {
              __typename: 'Human',
              name: 'default',
              father: null,
              nullFather: null,
              sons: [null],
            },
            {
              __typename: 'Human',
              name: 'default',
              father: null,
              nullFather: null,
              sons: [null],
            },
          ],
        },
      ],
    });
  });

  test('named fields object values - depth 1', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human(), ['father']);
    });

    expect(data).toEqual({
      father: {
        __typename: 'Human',
        name: 'default',
        father: null,
        nullFather: null,
        sons: [null],
      },
    });
  });

  test('named fields object values - depth 2', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human(), ['father'], 2);
    });

    expect(data).toEqual({
      father: {
        __typename: 'Human',
        name: 'default',
        father: {
          __typename: 'Human',
          name: 'default',
          father: null,
          nullFather: null,
          sons: [null],
        },
        nullFather: null,
        sons: [
          {
            __typename: 'Human',
            name: 'default',
            father: null,
            nullFather: null,
            sons: [null],
          },
          {
            __typename: 'Human',
            name: 'default',
            father: null,
            nullFather: null,
            sons: [null],
          },
        ],
      },
    });
  });

  test('named non-existent field', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.human(), ['non_existent_field']);
    });

    expect(data).toStrictEqual({});
  });

  test('null accessor', async () => {
    const { query, resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(query.nullArray);
    });

    expect(data).toBe(null);
  });

  test('primitive wrong accessor', async () => {
    const { resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields(123 as any);
    });

    expect(data).toBe(123);
  });

  test('object wrong accessor', async () => {
    const { resolved } = await createTestClient();

    const data = await resolved(() => {
      return selectFields({
        a: 1,
      });
    });

    expect(data).toStrictEqual({
      a: 1,
    });
  });
});

describe('refetch function', () => {
  test('refetch works', async () => {
    const { query, refetch, scheduler } = await createTestClient();

    const a = query.hello;

    expect(a).toBe(null);

    expect(scheduler.resolving).toBeTruthy();

    await scheduler.resolving!.promise;

    const a2 = query.hello;

    expect(scheduler.resolving).toBe(null);

    expect(a2).toBe('hello world');

    const a3Promise = refetch(() => query.hello);

    expect(scheduler.resolving).toBeTruthy();

    const a3 = await a3Promise;

    expect(a3).toBe(a2);
  });

  test('warns about no selections inside function, except on production', async () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation((message) => {
      expect(message).toBe('gqless: No selections made!');
    });
    const prevEnv = process.env.NODE_ENV;

    try {
      const { refetch } = await createTestClient();

      const value = await refetch(() => {
        return 123;
      });

      expect(value).toBe(123);
      expect(spy).toBeCalledTimes(1);

      process.env.NODE_ENV = 'production';

      const value2 = await refetch(() => {
        return 456;
      });

      expect(value2).toBe(456);
      expect(spy).toBeCalledTimes(1);
    } finally {
      process.env.NODE_ENV = prevEnv;
      spy.mockRestore();
    }
  });

  test('refetch proxy selections', async () => {
    const { query, resolved, refetch, cache } = await createTestClient();

    const time1 = await resolved(() => query.time);

    const cacheSnapshot1 = JSON.stringify(cache);

    expect(time1).toBeTruthy();

    const time2 = query.time;

    const cacheSnapshot2 = JSON.stringify(cache);

    expect(cacheSnapshot1).toBe(cacheSnapshot2);

    expect(time2).toBe(time1);

    await refetch(query);

    const cacheSnapshot3 = JSON.stringify(cache);

    expect(cacheSnapshot3).not.toBe(cacheSnapshot2);

    const time3 = query.time;

    const cacheSnapshot4 = JSON.stringify(cache);

    expect(cacheSnapshot4).toBe(cacheSnapshot3);

    expect(time3).not.toBe(time1);

    const noSelectionsToRefetchData = await refetch(query.nullArray);

    expect(noSelectionsToRefetchData).toBeTruthy();

    const cacheSnapshot5 = JSON.stringify(cache);

    expect(cacheSnapshot5).toBe(cacheSnapshot3);

    const hello = await resolved(() => query.hello);

    const cacheSnapshot6 = JSON.stringify(cache);

    expect(cacheSnapshot6).not.toBe(cacheSnapshot5);

    expect(hello).toBe('hello world');

    const time4 = query.time;

    expect(time4).toBe(time3);

    await refetch(query);

    const time5 = query.time;

    expect(time5).not.toBe(time4);
  });

  test('refetch proxy selections with invalid proxy', async () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation((message) => {
      expect(message).toBe('gqless: Invalid proxy to refetch!');
    });

    const prevNODE_ENV = process.env.NODE_ENV;
    try {
      const { refetch } = await createTestClient();

      const invalidProxy = {};
      const refetchData = await refetch(invalidProxy);

      expect(spy).toBeCalledTimes(1);
      expect(refetchData).toBe(invalidProxy);

      process.env.NODE_ENV = 'production';

      const refetchData2 = await refetch(invalidProxy);

      expect(spy).toBeCalledTimes(1);
      expect(refetchData2).toBe(invalidProxy);
    } finally {
      process.env.NODE_ENV = prevNODE_ENV;
      spy.mockRestore();
    }
  });
});
