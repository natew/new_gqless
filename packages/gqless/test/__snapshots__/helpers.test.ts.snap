// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectFields named fields array values - depth 1 1`] = `
Object {
  "sons": Array [
    Object {
      "__typename": "Human",
      "dogs": Array [
        null,
      ],
      "father": null,
      "id": "1",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        null,
      ],
    },
    Object {
      "__typename": "Human",
      "dogs": Array [
        null,
      ],
      "father": null,
      "id": "1",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        null,
      ],
    },
  ],
}
`;

exports[`selectFields named fields array values - depth 2 1`] = `
Object {
  "sons": Array [
    Object {
      "__typename": "Human",
      "dogs": Array [
        Object {
          "__typename": "Dog",
          "id": "1",
          "name": "a",
          "owner": null,
        },
        Object {
          "__typename": "Dog",
          "id": "2",
          "name": "b",
          "owner": null,
        },
      ],
      "father": Object {
        "__typename": "Human",
        "dogs": Array [
          null,
          null,
        ],
        "father": null,
        "id": "1",
        "name": "default",
        "nullFather": null,
        "sons": Array [
          null,
          null,
        ],
      },
      "id": "1",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        Object {
          "__typename": "Human",
          "dogs": Array [
            null,
            null,
          ],
          "father": null,
          "id": "1",
          "name": "default",
          "nullFather": null,
          "sons": Array [
            null,
            null,
          ],
        },
        Object {
          "__typename": "Human",
          "dogs": Array [
            null,
            null,
          ],
          "father": null,
          "id": "1",
          "name": "default",
          "nullFather": null,
          "sons": Array [
            null,
            null,
          ],
        },
      ],
    },
    Object {
      "__typename": "Human",
      "dogs": Array [
        Object {
          "__typename": "Dog",
          "id": "1",
          "name": "a",
          "owner": null,
        },
        Object {
          "__typename": "Dog",
          "id": "2",
          "name": "b",
          "owner": null,
        },
      ],
      "father": Object {
        "__typename": "Human",
        "dogs": Array [
          null,
          null,
        ],
        "father": null,
        "id": "1",
        "name": "default",
        "nullFather": null,
        "sons": Array [
          null,
          null,
        ],
      },
      "id": "1",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        Object {
          "__typename": "Human",
          "dogs": Array [
            null,
            null,
          ],
          "father": null,
          "id": "1",
          "name": "default",
          "nullFather": null,
          "sons": Array [
            null,
            null,
          ],
        },
        Object {
          "__typename": "Human",
          "dogs": Array [
            null,
            null,
          ],
          "father": null,
          "id": "1",
          "name": "default",
          "nullFather": null,
          "sons": Array [
            null,
            null,
          ],
        },
      ],
    },
  ],
}
`;

exports[`selectFields named fields object values - depth 1 1`] = `
Object {
  "father": Object {
    "__typename": "Human",
    "dogs": Array [
      null,
    ],
    "father": null,
    "id": "1",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      null,
    ],
  },
}
`;

exports[`selectFields named fields object values - depth 2 1`] = `
Object {
  "father": Object {
    "__typename": "Human",
    "dogs": Array [
      Object {
        "__typename": "Dog",
        "id": "1",
        "name": "a",
        "owner": null,
      },
      Object {
        "__typename": "Dog",
        "id": "2",
        "name": "b",
        "owner": null,
      },
    ],
    "father": Object {
      "__typename": "Human",
      "dogs": Array [
        null,
        null,
      ],
      "father": null,
      "id": "1",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        null,
        null,
      ],
    },
    "id": "1",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      Object {
        "__typename": "Human",
        "dogs": Array [
          null,
          null,
        ],
        "father": null,
        "id": "1",
        "name": "default",
        "nullFather": null,
        "sons": Array [
          null,
          null,
        ],
      },
      Object {
        "__typename": "Human",
        "dogs": Array [
          null,
          null,
        ],
        "father": null,
        "id": "1",
        "name": "default",
        "nullFather": null,
        "sons": Array [
          null,
          null,
        ],
      },
    ],
  },
}
`;

exports[`selectFields named no recursive 1`] = `
Object {
  "father": Object {
    "name": "default",
  },
  "name": "bar",
}
`;

exports[`selectFields named recursive - array 1`] = `
Array [
  Object {
    "name": "default",
  },
  Object {
    "name": "default",
  },
]
`;

exports[`selectFields named recursive, depth 1 1`] = `
Object {
  "father": Object {
    "__typename": "Human",
    "dogs": Array [
      null,
    ],
    "father": null,
    "id": "2",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      null,
    ],
  },
}
`;

exports[`selectFields named recursive, depth 2 1`] = `
Object {
  "father": Object {
    "__typename": "Human",
    "dogs": Array [
      Object {
        "__typename": "Dog",
        "id": "1",
        "name": "a",
        "owner": null,
      },
      Object {
        "__typename": "Dog",
        "id": "2",
        "name": "b",
        "owner": null,
      },
    ],
    "father": Object {
      "__typename": "Human",
      "dogs": Array [
        null,
        null,
      ],
      "father": null,
      "id": "2",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        null,
        null,
      ],
    },
    "id": "2",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      Object {
        "__typename": "Human",
        "dogs": Array [
          null,
          null,
        ],
        "father": null,
        "id": "2",
        "name": "default",
        "nullFather": null,
        "sons": Array [
          null,
          null,
        ],
      },
      Object {
        "__typename": "Human",
        "dogs": Array [
          null,
          null,
        ],
        "father": null,
        "id": "2",
        "name": "default",
        "nullFather": null,
        "sons": Array [
          null,
          null,
        ],
      },
    ],
  },
}
`;

exports[`selectFields recursive * - array 1`] = `
Array [
  Object {
    "__typename": "Human",
    "dogs": Array [
      null,
    ],
    "father": null,
    "id": "1",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      null,
    ],
  },
  Object {
    "__typename": "Human",
    "dogs": Array [
      null,
    ],
    "father": null,
    "id": "1",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      null,
    ],
  },
]
`;

exports[`selectFields recursive *, depth 1 1`] = `
Object {
  "__typename": "Human",
  "dogs": Array [
    null,
  ],
  "father": null,
  "id": "1",
  "name": "foo",
  "nullFather": null,
  "sons": Array [
    null,
  ],
}
`;

exports[`selectFields recursive *, depth 2 1`] = `
Object {
  "__typename": "Human",
  "dogs": Array [
    Object {
      "__typename": "Dog",
      "id": "1",
      "name": "a",
      "owner": null,
    },
    Object {
      "__typename": "Dog",
      "id": "2",
      "name": "b",
      "owner": null,
    },
  ],
  "father": Object {
    "__typename": "Human",
    "dogs": Array [
      null,
    ],
    "father": null,
    "id": "2",
    "name": "default",
    "nullFather": null,
    "sons": Array [
      null,
    ],
  },
  "id": "1",
  "name": "foo",
  "nullFather": null,
  "sons": Array [
    Object {
      "__typename": "Human",
      "dogs": Array [
        null,
      ],
      "father": null,
      "id": "2",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        null,
      ],
    },
    Object {
      "__typename": "Human",
      "dogs": Array [
        null,
      ],
      "father": null,
      "id": "2",
      "name": "default",
      "nullFather": null,
      "sons": Array [
        null,
      ],
    },
  ],
}
`;
